### PR TITLE
SCP-926: Use `iohk-monitoring` in SCB

### PR DESCRIPTION
**Pull request by j-mueller**
Use [`iohk-monitoring`](https://github.com/input-output-hk/iohk-monitoring-framework/tree/master/iohk-monitoring) to deal with the log outputs.

* `Plutus.SCB.Monitoring` has interpreters for the `LogMsg` and `LogObserve` effects that call the appropriate functions from `iohk-monitoring`
  * it also has some conveniences for dealing with `Configuration` values
* All logging types have `Cardano.BM.Data.Tracer.ToObject` instances. For types that are defined in `plutus-scb` itself, I put the instances right next to the types. For other types I put them in `Plutus.SCB.Instances`.
* The command-line interface to the app has changed slightly (cc @krisajenkins )
  * There is a new argument `--log-config` for the logging configuration YAML (optional)
    * The default config is used if there is no file provided
  * There is a new command `default-logging-config` that writes the default logging config YAML to a file
  * The `--verbose` flag sets the min. severity for logging to `Debug`, overriding the value that was configured in the logging config
  * `--ekg` is now a flag (used to be an argument with the EKG port). If it is set then EKG will run on the port that was configured in the logging configuration file.
* As far as the logging output itself is concerned, I left it mostly in the same state as before. The only exception is that `processAllContractOutboxes` now collects some basic timing and memory statistics (mostly to demonstrate how to do it). The type of metrics that are collected, and the functions they are collected for, can be configured in the logging config.

### CODE REVIEW

**Comment by j-mueller** [comment]
Just have some rebasing to do, but the reviewing may as well start now

File: cabal.project [hunk 1]
```diff
@@ -190,12 +194,13 @@ source-repository-package
 
 source-repository-package
   type: git
-  location: https://github.com/raduom/iohk-monitoring-framework
-  tag: b5c035ad4e226d634242ad5979fa677921181435
+  location: https://github.com/j-mueller/iohk-monitoring-framework
```
Comment (michaelpj) [inline cabal.project:197]: What's the issue here?

File: plutus-scb/app/Main.hs [hunk 2]
```diff
@@ -338,68 +378,78 @@ runCliCommand minLogLevel config serviceAvailability (ForkCommands commands) =
     forkCommand ::  Command -> IO (Async ())
     forkCommand subcommand = do
       putStrLn $ "Starting: " <> show subcommand
-      asyncId <- async . void . runApp minLogLevel config . renderLogMessages . runCliCommand minLogLevel config serviceAvailability $ subcommand
+      let trace' = natTracer (\x -> TraceLoggerT $ \_ -> x) trace
```
Comment (michaelpj) [inline plutus-scb/app/Main.hs:381]: Not entirely sure what's going on with things like this. Maybe we need a Note about the iohk-monitoring usage?

File: plutus-scb/app/Main.hs [hunk 3]
```diff
@@ -338,68 +378,78 @@ runCliCommand minLogLevel config serviceAvailability (ForkCommands commands) =
...
-    (minLogLevel, configPath, cmd) <-
+    (minLogLevel, configPath, logConfigPath, ekg, cmd) <-
         customExecParser
             (prefs $ disambiguate <> showHelpOnEmpty <> showHelpOnError)
             (info (helper <*> versionOption <*> commandLineParser) idm)
     config <- liftIO $ decodeFileThrow configPath
-    traverse_ (EKG.forkServer "localhost") (monitoringPort <$> monitoringConfig config)
+
+    logConfig <- maybe defaultConfig loadConfig logConfigPath
+    _ <- case minLogLevel of
```
Comment (michaelpj) [inline plutus-scb/app/Main.hs:434]: I discovered a nice idiom for this recently:
```
for_ minLogLevel $ \ll -> CM.setMinseverity logConfig ll
```

File: plutus-scb/app/Main.hs [hunk 4]
```diff
@@ -338,68 +378,78 @@ runCliCommand minLogLevel config serviceAvailability (ForkCommands commands) =
...
+            Nothing -> pure ()
+            Just ll -> CM.setMinSeverity logConfig ll
+    (trace :: Trace IO AppMsg, switchboard) <- setupTrace_ logConfig "pab"
+    let trace' = natTracer (\x -> TraceLoggerT $ \_ -> x) trace
```
Comment (michaelpj) [inline plutus-scb/app/Main.hs:438]: Some duplication with above, perhaps?

File: plutus-scb/app/Main.hs [hunk 5]
```diff
@@ -338,68 +378,78 @@ runCliCommand minLogLevel config serviceAvailability (ForkCommands commands) =
...
             (prefs $ disambiguate <> showHelpOnEmpty <> showHelpOnError)
             (info (helper <*> versionOption <*> commandLineParser) idm)
     config <- liftIO $ decodeFileThrow configPath
-    traverse_ (EKG.forkServer "localhost") (monitoringPort <$> monitoringConfig config)
+
+    logConfig <- maybe defaultConfig loadConfig logConfigPath
+    _ <- case minLogLevel of
+            Nothing -> pure ()
+            Just ll -> CM.setMinSeverity logConfig ll
+    (trace :: Trace IO AppMsg, switchboard) <- setupTrace_ logConfig "pab"
```
Comment (michaelpj) [inline plutus-scb/app/Main.hs:437]: Seems like a possible important magic string -> top level constant.

File: plutus-scb/app/Main.hs [hunk 6]
```diff
@@ -338,68 +378,78 @@ runCliCommand minLogLevel config serviceAvailability (ForkCommands commands) =
...
+    let trace' = natTracer (\x -> TraceLoggerT $ \_ -> x) trace
+
+    -- enable EKG backend
+    when (ekg == YesEKGServer) $ do
```
Comment (michaelpj) [inline plutus-scb/app/Main.hs:441]: do block with a single line is redundant

File: plutus-scb/app/Main.hs [hunk 7]
```diff
@@ -338,68 +378,78 @@ runCliCommand minLogLevel config serviceAvailability (ForkCommands commands) =
...
+        Cardano.BM.Backend.EKGView.plugin logConfig trace switchboard >>= loadPlugin switchboard
+
     serviceAvailability <- newToken
     result <-
-        runApp minLogLevel config $ do
-            logInfo $ "Running: " <> Text.pack (show cmd)
-            renderLogMessages $ runCliCommand minLogLevel config serviceAvailability cmd
+        runApp (mapSCBMsg trace) logConfig config $ handleLogMsgTrace trace' $ do
```
Comment (michaelpj) [inline plutus-scb/app/Main.hs:446]: ditto

File: plutus-scb/src/Plutus/SCB/App.hs [hunk 8]
```diff
@@ -113,23 +114,25 @@ type AppBackend m =
...
     . runReader dbConnection
-    . runStderrLog
-    . handleObserveLog
-    . renderLogMessages
-    . renderLogMessages
-    . renderLogMessages
-    . renderLogMessages
-    . renderLogMessages
+    . handleObserveTrace config trace
+    . handleLogMsgTraceMap SCoreMsg trace
```
Comment (michaelpj) [inline plutus-scb/src/Plutus/SCB/App.hs:131]: I won't argue that this should take prisms ;)

File: plutus-scb/src/Plutus/SCB/App.hs [hunk 9]
```diff
@@ -168,57 +171,26 @@ runAppBackend e@Env{dbConnection, nodeClientEnv, walletClientEnv, signingProcess
...
             . WalletClient.handleWalletClient walletClientEnv
 
+type App a = Eff (AppBackend (TraceLoggerT IO)) a
 
-type App a = Eff (AppBackend (LoggingT IO)) a
-
-runApp :: LogLevel -> Config -> App a -> IO (Either SCBError a)
-runApp minLogLevel Config {dbConfig, nodeServerConfig, walletServerConfig, signingProcessConfig, chainIndexConfig} action =
-    runStdoutLoggingT $ filterLogger (\_ logLevel -> logLevel >= minLogLevel) $ do
+runApp :: Trace IO SCBLogMsg -> CM.Configuration -> Config -> App a -> IO (Either SCBError a)
```
Comment (michaelpj) [inline plutus-scb/src/Plutus/SCB/App.hs:176]: Nit: the trace/config arguments are in a different order to above.

File: plutus-scb/src/Plutus/SCB/Core.hs [hunk 10]
```diff
@@ -92,6 +101,20 @@ instance Pretty t => Pretty (CoreMsg t) where
...
+        Installing t ->
+            mkObjectStr "installing contract" [con t]
+        Installed ->
+            mkObjectStr "contract installed" []
+        FindingContract instanceID ->
+            mkObjectStr "finding contract instance" [inst instanceID]
+        FoundContract state ->
+            mkObjectStr "found contract" $
+                case v of
+                    MaximalVerbosity -> [("contrat_state", toJSON state)]
```
Comment (michaelpj) [inline plutus-scb/src/Plutus/SCB/Core.hs:115]:
```suggestion
                    MaximalVerbosity -> [("contract_state", toJSON state)]
```

File: plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import           Data.Text                             (Text)
+import           Data.Text.Prettyprint.Doc             (Pretty (..), defaultLayoutOptions, layoutPretty)
+import qualified Data.Text.Prettyprint.Doc.Render.Text as Render
+import           Plutus.SCB.Events.Contract            (ContractInstanceId, IterationID)
+
+mkObjectStr :: Text -> [(Text, Value)] -> HashMap Text Value
+mkObjectStr str rest =
+    mkObject $ ("string", String str) : rest
+
+inst :: ContractInstanceId -> (Text, Value)
```
Comment (michaelpj) [inline plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs:23]: It seems weird to me that this isn't somehow handled generically? Can't we just derive instances that have the right field names?

File: plutus-scb/src/Plutus/SCB/Monitoring.hs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  , MonadIO m
+  , MonadCatch m
+  )
+  => CM.Configuration
+  -> Trace m a
+  -> Eff (LogObserve (L.LogMessage Text) ': effs)
+  ~> Eff effs
+handleObserveTrace config t =
+
+  let observeBefore :: (L.LogMessage Text) -> Eff effs (Maybe (SubTrace, CounterState))
```
Comment (michaelpj) [inline plutus-scb/src/Plutus/SCB/Monitoring.hs:122]: This all seems a bit tricky, could do with some comments.

File: plutus-scb/src/Plutus/SCB/ParseStringifiedJSON.hs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    deriving anyclass (ToJSON, FromJSON)
+    deriving ToObject via (PrettyToObject UnStringifyJSONLog)
+
+instance Pretty UnStringifyJSONLog where
+    pretty = \case
+        ParseStringifiedJSONAttempt -> "parseStringifiedJSON: Attempting to remove 1 layer StringifyJSON"
+        ParseStringifiedJSONFailed -> "parseStringifiedJSON: Failed, returning original string"
+        ParseStringifiedJSONSuccess -> "parseStringifiedJSON: Succeeded"
+
+parseStringifiedJSON ::
```
Comment (michaelpj) [inline plutus-scb/src/Plutus/SCB/ParseStringifiedJSON.hs:39]: Why do we actually have this?

File: cabal.project (hunk 1 above)
Comment (j-mueller) [reply cabal.project:197]: Needed to increase the bounds on `base` to build with newer GHC. I can make a PR into @raduom 's fork .

File: plutus-scb/app/Main.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-scb/app/Main.hs:417]: Added a note in https://github.com/input-output-hk/plutus/pull/2223/commits/7e97a3075ab9017813cb1abbc1da8159e71550ee

File: plutus-scb/app/Main.hs (hunk 3 above)
Comment (j-mueller) [reply plutus-scb/app/Main.hs:470]: ah yes I've seen that one before, thanks.

File: plutus-scb/src/Plutus/SCB/App.hs (hunk 8 above)
Comment (j-mueller) [reply plutus-scb/src/Plutus/SCB/App.hs:131]: Then I won't respond by pointing out the fact that functions are more general than prisms and I am merely applying the principle of "be liberal in what you accept and strict in what you produce" ;)

File: plutus-scb/src/Plutus/SCB/ParseStringifiedJSON.hs (hunk 13 above)
Comment (j-mueller) [reply plutus-scb/src/Plutus/SCB/ParseStringifiedJSON.hs:39]: Some of the values that we get from the frontend contain string-encoded JSON values... (at least that was the case before the demo at the virtual summit, when I first added `parseStringifiedJSON`. I'll check if it is still needed)

File: plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs (hunk 11 above)
Comment (j-mueller) [reply plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs:23]: there doesn't seem to be a suitable class for this at the moment but I can try to write one

File: plutus-scb/src/Plutus/SCB/Monitoring.hs (hunk 12 above)
Comment (j-mueller) [reply plutus-scb/src/Plutus/SCB/Monitoring.hs:129]: done.

File: plutus-scb/app/Main.hs (hunk 4 above)
Comment (j-mueller) [reply plutus-scb/app/Main.hs]: Fixed it in f4be403f11f04c7eced9a137ad1c24360cdcaa5c

**Comment by j-mueller** [comment]
Two things left to do, (1) check if `ParseStringifyJSON` is still needed, (2) generic approach to creating log objects in `plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs `

File: cabal.project (hunk 1 above)
Comment (raduom) [reply cabal.project:193]: It's fine. I don't care much for maintaining it. :)

File: plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs (hunk 11 above)
Comment (j-mueller) [reply plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs]: I made `Cardano.BM.Data.Tracer.Extras` a bit more generic in c434c00c581121d90a16658e6c56c9914cb8fafa.

Comment (michaelpj) [reply plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs]: Looks reasonable - can we also ask upstream whether they have any tricks for this or would like a patch? Not sure quite who to ask, but I'm sure we can find someone.

Comment (j-mueller) [reply plutus-scb/src/Cardano/BM/Data/Tracer/Extras.hs]: https://github.com/input-output-hk/iohk-monitoring-framework/issues/580

**Comment by j-mueller** [comment]
I'll merge this in a little while, don't think it'll get any better.

**Review by michaelpj — APPROVED** [review]

**Pull request merged**